runtime/list.go

```diff
@@ -291,6 +291,37 @@ func listNE(f *Frame, v, w *Object) (*Object, *BaseException) {
 	return listCompare(f, toListUnsafe(v), w, NE)
 }
 
+func listPop(f *Frame, args Args, kwargs KWArgs) (*Object, *BaseException) {
+	argc := len(args)
+	expectedTypes := []*Type{ListType, ObjectType}
+	if argc == 1 {
+		expectedTypes = expectedTypes[:1]
+	}
+	if raised := checkMethodArgs(f, "pop", args, expectedTypes...); raised != nil {
+		return nil, raised
+	}
+	i := -1
+	if argc == 2 {
+		o, raised := IntType.Call(f, Args{args[1]}, nil)
+		if raised != nil {
+			return nil, raised
+		}
+		i = toIntUnsafe(o).Value()
+	}
+	l := toListUnsafe(args[0])
+	if i < 0 {
+		i += len(l.elems)
+	}
+	if i >= len(l.elems) || i < 0 {
+		return nil, f.RaiseType(IndexErrorType, "list index out of range")
+	}
+	l.mutex.Lock()
```

Comment:
You must unlock before returning. That's why you were getting a deadlock with the other code. But this code is still racy. The whole operation including checking the length needs to be atomic, something like:

```
l.mutex.Lock()
numElems := len(l.elems)
if i < 0 {
  i += numElems
}
var item *Object
var raised *BaseException
if i >= numElems || i < 0 {
  raised = f.RaiseType(IndexErrorType, "list index out of range")
} else {
  item = l.elems[i]
  l.elems = append(l.elems[:i], l.elems[i+1:]...)
}
l.mutex.Unlock()
return item, raised
```